### DISCUSSION TITLE
Passing request body from json file to artillery api tests.

### DISCUSSION THREAD

**Discussion by BhavnaSoni1991** [artilleryio/artillery#2638] [Artillery]
I want to add performance tests for api having quite complex request payload and hence want to pass them through json file. Does artillery support reading request payload from json file? In the examples I could see request payload values are directly passed in yml file and not read from any file.

**Comment by bernardobridge** [comment]
Hi @BhavnaSoni1991 👋 

IMO the easiest way to do it is to use a custom javascript function, where you can require the JSON file and then set the body directly to it:

`processor.js`
```javascript
const myJson = require('./path-to-json');

function setJSONBody(requestParams, context, ee, next) {
  requestParams.body = JSON.stringify(myJson);

  return next();
}
module.exports = { setJSONBody };
```

Then, you can call that in a [`beforeRequest`](https://www.artillery.io/docs/reference/engines/http#beforerequest-hooks) (or `beforeScenario`) hook.

```yaml
config:
    ...
    processor: './processor.js'
scenarios:
  - flow:
      - post:
          url: '/'
          beforeRequest: "setJSONBody"
```

**Reply by BhavnaSoni1991** [reply]
Thanks @bernardobridge the solution worked!!

**Comment by BhavnaSoni1991** [comment]
Hi @bernardobridge my script ran file on local using the code snippet above but in the gitlab pipeline I am getting following "MODULE_NOT_FOUND" error when trying to run on aws fargate.
I suspect that this is related to require module. Can you please help?

**Reply by bernardobridge** [reply]
Hi @BhavnaSoni1991 , I would need more details: 

- actual full error stack you saw
- your file structure
- did the list of files that Fargate lists in the beginning include your json?

**Reply by BhavnaSoni1991** [reply]
@bernardobridge I figured out that my json file is not being listed in AWS fargate Test bundle contents. I have kept this json file under resources folder. Do I need to do anything explicitly so that fargate lists it?

**Reply by bernardobridge** [reply]
Hi @BhavnaSoni1991 , the `json` file should be automatically detected by the Test bundle, so that might be a bug. Unfortunately, without more information, I cannot debug this. Can you please provide:

- Your full file structure (your scenario, config, processor and json, and their locations)
- The contents of your test yaml, and the processor file

Thanks!

**Comment by BhavnaSoni1991** [comment]
hi @bernardobridge below is the full file structure of my scenario file(test3.yml), config is same yml file (test3.yml), processor (helper.js), json (test3.json) file.
![image](https://github.com/artilleryio/artillery/assets/160220723/81bd5ec3-a26e-43a3-9b72-801cfb141ff5)

Below are the contents of yaml file and processor file
![image](https://github.com/artilleryio/artillery/assets/160220723/c95f7820-a342-471d-9239-5ca01ee7ce5b)

![image](https://github.com/artilleryio/artillery/assets/160220723/45f5e3b1-a9de-44c6-a316-9a2355577106)
Could you please have a look and let me know what is wrong?

**Reply by bernardobridge** [reply]
Hi @BhavnaSoni1991 .

The problem is the dynamic require on line 131. The code that bundles dependencies won't be able to resolve that to an actual file, as those variables only exist when the test is running.

Either use static requires (e.g. `require('./path_to_json.json')`), or use `includeFiles` in your config (with a list of all files you want Fargate to bundle):

```yaml
  config:
    ...
    includeFiles:
        - ./path_to_json.json
```

**Reply by BhavnaSoni1991** [reply]
Thanks @bernardobridge but even if I use includeFiles in the yml file I want the json file in my setJSONBody method to be dynamic which I need to pass from yml file. Is there a way to achieve it?

**Reply by bernardobridge** [reply]
Any file you add in `includeFiles` will be bundled and available in Fargate, so then you should be able to use the dynamic require in the setJSONBody.

**Reply by BhavnaSoni1991** [reply]
hi @bernardobridge does the path have to be given differently in includeFiles since I am trying to pass path same way I have done for processor and path below. I can now see the fargate is bundling the json but it is taking incorrect path?
![image](https://github.com/artilleryio/artillery/assets/160220723/9a11df4a-f8d5-4540-ac8c-3431acd22ba8)

**Reply by bernardobridge** [reply]
The `requestPath` you've set in this case is relative to the processor file, since you are requiring it dynamically from the processor file.

The path you set in `includeFiles` needs to be relative to the config.

**Reply by BhavnaSoni1991** [reply]
Thank you so much @bernardobridge for your quick response it worked !!

**Reply by BhavnaSoni1991** [reply]
hi @bernardobridge until now the solution described above worked for single virtual user. Now I am trying to execute the same for multiple virtual users which means I have to change the unique value in request payload for each virtual user. Since as of now I am setting request body using setJSONBody method as described above to which I was passing json file path and request key. Now in addition to that I also need to change some value of request payload? Could you please help me how can I achieve it using setJSONBody method?

**Reply by roydekleijn** [reply]
Have you found a solution for this?
